### PR TITLE
Yaksh chat

### PR DESCRIPTION
**Pull request by adityacp**
Chat system for Yaksh.
Supports Broadcast chatting between moderator and students.
Closes issue #285

### CODE REVIEW

File: yaksh/consumers.py [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+
+    if set(data.keys()) != set(('sender_id', 'message')):
+        message.reply_channel.send({'close': True})
+        return
+
+    if data:
+        context = {}
+        chat_dict = {}
+        message = room.messages.create(**data)
+        sender_tz = User.objects.get(id=data['sender_id']).profile.timezone
```
Comment (maheshgudi) [inline yaksh/consumers.py]: The timezone while saving should be saved as GMT. And while rendering the time, the time should be rendered according to the user's timezone. This doesn't happen currently

File: yaksh/static/yaksh/js/chat.js [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+    var chatsock;
+    var label;
+    var timer;
+    var page_title = document.title;
+    var course_id = $("#course_id").val();
+
+    $("#main").on("click", function(){
+        // Open chat window
+        var room_url = window.location.protocol + "//" +
+                    window.location.host + "/exam/new/chat/" + course_id;
```
Comment (maheshgudi) [inline yaksh/static/yaksh/js/chat.js]: I am not sure if the <code>room_url</code> should be a part of the client side. Or at least it should encrypted in some form. Exposing the url to the client can increase vulnerability.

File: yaksh/static/yaksh/js/chat.js [hunk 3]
```diff
@@ -0,0 +1,166 @@
...
+        ajax_request(room_url);
+    });
+
+    function ajax_request(chat_url) {
+        // Get messages from ajax request
+        $.ajax({
+            url: chat_url,
+            timeout:30000,
```
Comment (maheshgudi) [inline yaksh/static/yaksh/js/chat.js]: 30000 seconds?

File: yaksh/static/yaksh/js/chat.js [hunk 4]
```diff
@@ -0,0 +1,166 @@
...
+                message: $('#message').html(),
+            }
+            chatsock.send(JSON.stringify(message));
+        }
+        $("#message").empty();
+        return false;
+    });
+
+    // Create divs for showing sent and received messages
+    var div_sent = "<div class='row msg_container base_sent'>\n"+
```
Comment (maheshgudi) [inline yaksh/static/yaksh/js/chat.js]: Cant this be done with <code>document.createElement</code>? This is just horrible!

File: yaksh/static/yaksh/js/chat.js (hunk 2 above)
Comment (adityacp) [reply yaksh/static/yaksh/js/chat.js]: I feel it is okay. Just need to make sure from views that no intruder can get messages. I feel I covered checks for user to get messages in views function

**Comment by ankitjavalkar** [comment]
Need to add feature to changelog

**Comment by prathamesh920** [comment]
Closing for now.
Will reopen with 0.8 release.

**Pull request closed without merging**